### DISCUSSION TITLE
Confusion about tags

### DISCUSSION THREAD

**Discussion by Antonytm** [microsoft/dotnet-framework-docker#965] [Q&A] [answered]
I look on page https://hub.docker.com/_/microsoft-dotnet-framework-sdk/

I see that Windows Server Core 2022 build has tags  `4.8, latest`
I see that Windows Server Core, version 20H2 build has tags  `4.8, latest`
I see that Windows Server Core 2019 build has tags  `4.8, latest`
I see that Windows Server Core 2016 build has tags  `4.8, latest`

As many versions have the same tag, I am confused, about what version will be used, when I will write:
```
docker pull mcr.microsoft.com/dotnet/framework/sdk:4.8 
```
?

**Comment by richlander** [comment]
It depends on your Windows (Client or Server) version. I am not sure if that is documented anywhere.

You can pull the image and then check which Windows Server Core version you pulled. Does that work?

**Comment by Antonytm** [comment]
I have Windows 10 21H1 locally, I pull `mcr.microsoft.com/dotnet/framework/sdk:4.8`, I got `Digest: sha256:296a507e6fe9cadd75a31a8781fccaf747447d66bcc51463abb705f8cab7b60b`. 

Then I repeat the same in Github actions using `windows-latest`, which is `windows server 2022`. I checked digest, it is the same `296a507e6fe9cadd75a31a8781fccaf747447d66bcc51463abb705f8cab7b60b`.

@richlander Should I get different versions?

**Comment by mthalman** [comment, marked as answer]
The `4.8` is a manifest list (multi-platform) tag. Docker will automatically resolve that tag to an actual image based on rules governed by the metadata of the manifest list. For Windows, those rules make use of the Windows version. It will attempt to find an image that matches the Windows version of the host.

If we take a look at manifest list, we can see what we're dealing with:

```console
C:\> docker manifest inspect mcr.microsoft.com/dotnet/framework/sdk:4.8
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 3206,
         "digest": "sha256:fb80acb2729830be7b7e7f2a9be061a9ffa72f33a2fbe9fc75aed84f46857088",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.14393.5192"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2788,
         "digest": "sha256:82e0a5e2c8747b9f2c88a15e7f652fd4b437cbe0399ab15ac54a0680dde8b453",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.17763.3046"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2580,
         "digest": "sha256:e93be4dcc24e1327ec07e843c5b391e7c84dc9897b2deebb813b06f3547c0745",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.19042.1766"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2579,
         "digest": "sha256:6c490f4528cb625c60aa119669830639ae3988c3c23dfb2f94a4ef19df2eb606",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.20348.768"
         }
      }
   ]
}
```

That metadata provides a mapping between image digests and Windows version. So there's a registration for digest `sha256:6c490f4528cb625c60aa119669830639ae3988c3c23dfb2f94a4ef19df2eb606` that is based on Windows Server 2022 (10.0.20348.768). So if you pulled the `4.8` tag on a Windows Server 2022 machine, that's the image you'd get. 

Windows desktop versions are more problematic now that Windows Server has moved to releasing Windows Server versions only for LTSC (long term support channel) versions and not SAC (semi-annual channel). This means your Windows 10 21H1 (version 10.0.19043) machine doesn't have a matching Windows Server container version. This is true for all of Windows Server containers, not specifically .NET Framework containers. Because no match is found, it falls back to the first version in the manifest list, version 10.0.14393.5192, which is Windows Server 2016.

If you want to be explicit about which version you get on your desktop machine, you'd need to use a tag which specifies the Windows version. For example, if you want the Windows Server 2022 image, you'd need to use the `4.8-windowsservercore-ltsc2022` tag. This kind of workaround is only really needed for desktop scenarios. When running from a Windows Server machine, you'll always get a direct matching version when targeting just the `4.8` tag.

One trick you could do with your Dockerfile is to make use of an `ARG` that allows you to override the tag for desktop scenarios:

```Dockerfile
ARG NET_FX_TAG=4.8
FROM mcr.microsoft.com/dotnet/framework/sdk:$NET_FX_TAG
```

To use this from the command line you would use the `--build-arg` option: `docker build --build-arg NET_FX_TAG=4.8-windowsservercore-ltsc2022 -t my-tag .`


Side note regarding the digest you had provided:

Digest `296a507e6fe9cadd75a31a8781fccaf747447d66bcc51463abb705f8cab7b60b` refers to the `sdk:4.8` manifest list. It doesn't refer to the underlying image that actually gets resolved and pulled down. So that digest doesn't help indicate what image was pulled.

**Comment by Antonytm** [comment]
@mthalman 

Thank you! That is a very good explanation!

Is there any quick way to compare configurations for different platforms?

I had an issue with differences in the default shell on `windows-2019` and `windows-2022`
I know that the default shell on Windows 10 is PowerShell by running the command:
`docker inspect mcr.microsoft.com/dotnet/framework/sdk:4.8`
```
            "Cmd": [
                "powershell",
                "-Command",
                "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';",
                "#(nop) ",
                "SHELL [powershell -Command $ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';]"
            ],
```

And for `windows-2019` it is also PoserShell.
But for `windows-2022`, it is cmd.

But I am not even able to run locally `docker inspect mcr.microsoft.com/dotnet/framework/sdk:4.8-20220614-windowsservercore-ltsc2022` and compare because I am on Windows 10. And to run it on a server I needed to add this step into CI.

**Reply by mthalman** [reply]
> Is there any quick way to compare configurations for different platforms?

The simplest way would probably be to just use a file comparison tool to compare the Dockerfiles that are in this repo. So you could compare [`sdk:4.8-windowsservercore-ltsc2019`](https://github.com/microsoft/dotnet-framework-docker/blob/main/src/sdk/4.8/windowsservercore-ltsc2019/Dockerfile) to [`sdk:4.8-windowsservercore-ltsc2022`](https://github.com/microsoft/dotnet-framework-docker/blob/main/src/sdk/4.8/windowsservercore-ltsc2022/Dockerfile).

> I had an issue with differences in the default shell on windows-2019 and windows-2022

Yes, setting the shell to PowerShell was something that specifically existed in the WS 2019 SDK Dockerfile. It was then later changed as part of the release of Windows Server 20H2 and documented at that time: https://github.com/microsoft/dotnet-framework-docker/discussions/925.
